### PR TITLE
update MoE PoC features & bfloat16 model load

### PR DESCRIPTION
**Pull request by rebel-wonsubkim**
MoE PoC update PR
- implement  custom DP multicast based on torch.distributed.all_gather
- include LMHead logits all_gather into model execution
- support custom MoE block using moe_custom_kernel & masked input
- enable model config for bf16 data type
- support torch.distributed process group for ccl subgroup
- implement custom device communicator including all_gather

current status
- qwen1.5-moe model checked - tp4/ep4, dp2+tp2/ep4, pp2+tp2/ep2
- qwen3-mode model checked - tp8/ep8, pp2+tp4/ep4, dp2+tp4/ep8 (oom)

### CODE REVIEW

**Comment by rebel-jiwoopark** [comment]
@rebel-jaehwang @huijjj PTAL

File: vllm_rbln/rbln_envs.py [hunk 1]
```diff
@@ -46,7 +46,7 @@
      ("true", "1")),
     # Enable warmup
     "VLLM_RBLN_ENABLE_WARM_UP":
-    (lambda: os.environ.get("VLLM_RBLN_ENABLE_WARM_UP", "True").lower() in
+    (lambda: os.environ.get("VLLM_RBLN_ENABLE_WARM_UP", "False").lower() in
```
Comment (rebel-jonghewk) [inline vllm_rbln/rbln_envs.py]: Is there a reason for changing the default behavior? For optimum-rbln based, we are using warmp up for our sampler.

File: vllm_rbln/rbln_communicator.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
+# Copyright 2025 Rebellions Inc. All rights reserved.
```
Comment (rebel-jonghewk) [inline vllm_rbln/rbln_communicator.py]: How about adding it inside a new directory, (e.g. vllm_rbln/distributed/...) ?

File: examples/model_test_compare.py [hunk 3]
```diff
@@ -0,0 +1,483 @@
+#!/usr/bin/env python3
```
Comment (rebel-jonghewk) [inline examples/model_test_compare.py]: I don't think we need this anymore. You can delete it.

File: vllm_rbln/rbln_envs.py [hunk 4]
```diff
@@ -57,6 +57,28 @@
     "VLLM_RBLN_FLASH_CAUSAL_ATTN":
     (lambda: os.environ.get("VLLM_RBLN_FLASH_CAUSAL_ATTN", "True").lower() in
      ("true", "1")),
+    # enforce model data type into fp32 not model_config.dtype
+    "RBLN_ENFORCE_MODEL_FP32":
```
[the remark this answers is not in the document]
Comment (rebel-wonsubkim) [reply vllm_rbln/rbln_envs.py]: will follow this rule

File: vllm_rbln/rbln_envs.py (hunk 1 above)
Comment (rebel-wonsubkim) [reply vllm_rbln/rbln_envs.py]: just for PP, enabled warm-up but if there is conflict case, I will fix it

File: vllm_rbln/distributed/rbln_communicator.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
+# Copyright 2025 Rebellions Inc. All rights reserved.
```
Comment (rebel-wonsubkim) [reply vllm_rbln/distributed/rbln_communicator.py]: okay got it

**Comment by rebel-jonghewk** [comment]
@rebel-eunji Could you add the necessary dtype conversions when the input is bf16 but the rbln_config dtype differs?

File: vllm_rbln/rbln_envs.py (hunk 1 above)
Comment (rebel-jonghewk) [reply vllm_rbln/rbln_envs.py]: @rebel-wonsubkim Can you revert this back to true as default?

File: vllm_rbln/distributed/rbln_communicator.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+                                                  (self.world_size *
+                                                   input_size[dim], ) +
+                                                  input_size[dim + 1:])
+        elif dim == 0:
+            pass
+        else:
+            raise AssertionError("RBLN all_gather dim!=0, dim!=-1, NYI")
+            # Reshape
+            output_tensor = output_tensor.reshape((self.world_size, ) +
+                                                  input_size)
```
[the remark this answers is not in the document]
Comment (rebel-jiwoopark) [reply vllm_rbln/distributed/rbln_communicator.py]: Thank you for reply
